concordium-consensus/src/Concordium/GlobalState/Persistent/LMDB.hs

```diff
@@ -199,30 +275,124 @@ makeDatabaseHandlers treeStateDir readOnly = do
-initializeDatabase gb stRef rp@RuntimeParameters{..} = do
-  -- The initial mapsize needs to be high enough to allocate the genesis block and its finalization record or
-  -- initialization would fail. It also needs to be a multiple of the OS page size. We considered keeping 4096 as a typical
-  -- OS page size and setting the initial mapsize to 64MB which is very unlikely to be reached just by the genesis block.
-  createDirectoryIfMissing False rpTreeStateDir
-  handlers@DatabaseHandlers{..} <- databaseHandlers rp
-  let gbh = getHash gb
-      gbfin = FinalizationRecord 0 gbh emptyFinalizationProof 0
-  transaction _storeEnv False $ \txn -> do
-    storeRecord txn _finalizedByHeightStore 0 gbh
-    storeRecord txn _blockStore gbh StoredBlock {
+initializeDatabase :: forall pv st ati bs. (IsProtocolVersion pv, S.Serialize st) =>
+  -- |Genesis block pointer
+  PersistentBlockPointer pv ati bs ->
+  -- |Genesis block state
+  st ->
+  -- |Tree state directory
+  FilePath ->
+  IO (DatabaseHandlers pv st)
+initializeDatabase gb stRef treeStateDir = do
+  createDirectoryIfMissing False treeStateDir
+  let storedGenesis = StoredBlock {
                     sbFinalizationIndex = 0,
                     sbInfo = _bpInfo gb,
                     sbBlock = _bpBlock gb,
                     sbState = stRef
                   }
+  -- The initial mapsize needs to be high enough to allocate the genesis block and its finalization record or
+  -- initialization would fail. Since a regenesis block can contain a serialization of the state, which may be
+  -- arbitrarily large, to be safe we ensure that we have at least 1MB more than the size of the serialization
+  -- of the genesis block.
+  let initSize = fromIntegral $ LBS.length (S.encodeLazy storedGenesis) + 1_048_576
+  handlers@DatabaseHandlers{..} <- makeDatabaseHandlers treeStateDir False initSize
+  let gbh = getHash gb
+      gbfin = FinalizationRecord 0 gbh emptyFinalizationProof 0
+  transaction _storeEnv False $ \txn -> do
+    storeRecord txn _finalizedByHeightStore 0 gbh
+    storeRecord txn _blockStore gbh storedGenesis
     storeRecord txn _finalizedByHeightStore 0 gbh
     storeRecord txn _finalizationRecordStore 0 gbfin
+    storeRecord txn _metadataStore versionMetadata $
+        S.encode $ VersionMetadata {
+            vmDatabaseVersion = 0,
+            vmProtocolVersion = demoteProtocolVersion (protocolVersion @pv)
+        }
   return handlers
 
+-- |Add a database version record to a database (if one is not already present).
+-- The record indicates database version 0 and protocol version 'P1', which is appropriate when
+-- migrating a database from an earlier version.
+addDatabaseVersion :: (MonadLogger m, MonadIO m) => FilePath -> m ()
+addDatabaseVersion treeStateDir = do
+  handlers :: DatabaseHandlers 'P1 () <- liftIO $ makeDatabaseHandlers treeStateDir False dbInitSize
+  handlers' <- execStateT
+    (resizeOnFull 4096 $ \h -> transaction (_storeEnv h) False $ \txn -> 
```

Comment:
Why was 4096 chosen? Just because it is the default page size? Could you clarify for future reference.